src/EnergyPlus/WaterThermalTanks.cc

```diff
@@ -6479,64 +6483,69 @@ namespace WaterThermalTanks {
-				Qmix = InvMixUp * Cp * ( TempUp - NodeTemp ) + InvMixDn * Cp * ( TempDn - NodeTemp );
+				Qmix = Cp * ( InvMixUp * delTempUp + InvMixDn * delTempDn );
 
 				if ( Tank.HeaterOn1 || Tank.HeaterOn2 ) {
-					LossCoeff = Tank.Node( NodeNum ).OnCycLossCoeff;
+					LossCoeff = tank_node.OnCycLossCoeff;
 					Qloss = LossCoeff * ( AmbientTemp - NodeTemp );
-					Qlosszone = Qloss * Tank.SkinLossFracToZone;
-					Qoncycheat = Tank.Node( NodeNum ).OnCycParaLoad * Tank.OnCycParaFracToTank;
+					Qheat = Qoncycheat = tank_node.OnCycParaLoad * Tank.OnCycParaFracToTank;
+					Qusl = Quse + Qsource + Qloss;
 
-					Qneeded = max( -Quse - Qsource - Qloss - Qoncycheat, 0.0 );
+					Qneeded = max( -Qusl - Qoncycheat, 0.0 );
 
-					Qheat = Qoncycheat;
-					if ( NodeNum == Tank.HeaterNode1 ) Qheat += Qheater1;
-					if ( NodeNum == Tank.HeaterNode2 ) Qheat += Qheater2;
+					if ( NodeNum == Tank.HeaterNode1 ) {
+						Qheat += Qheater1;
+					} else if ( NodeNum == Tank.HeaterNode2 ) {
+						Qheat += Qheater2;
+					}
 				} else {
-					LossCoeff = Tank.Node( NodeNum ).OffCycLossCoeff;
+					LossCoeff = tank_node.OffCycLossCoeff;
 					Qloss = LossCoeff * ( AmbientTemp - NodeTemp );
-					Qlosszone = Qloss * Tank.SkinLossFracToZone;
-					Qoffcycheat = Tank.Node( NodeNum ).OffCycParaLoad * Tank.OffCycParaFracToTank;
+					Qheat = Qoffcycheat = tank_node.OffCycParaLoad * Tank.OffCycParaFracToTank;
+					Qusl = Quse + Qsource + Qloss;
 
-					Qneeded = max( -Quse - Qsource - Qloss - Qoffcycheat, 0.0 );
-					Qheat = Qoffcycheat;
+					Qneeded = max( -Qusl - Qoffcycheat, 0.0 );
 				}
 
-				Qunmet = max( Qneeded - Qheater1 - Qheater2, 0.0 );
+				Qunmet = max( Qneeded - Qheaters, 0.0 );
 
 				// Calculate node heat balance
-				Tank.Node( NodeNum ).NewTemp = NodeTemp + ( Quse + Qsource + Qcond + Qflow + Qmix + Qloss + Qheat ) * dt / ( NodeMass * Cp );
+				tank_node.NewTemp = NodeTemp + ( Qusl + Qcond + Qflow + Qmix + Qheat ) * dt_Cp / NodeMass;
 
 				if ( ! Tank.IsChilledWaterTank ) {
-					if ( ( NodeNum == 1 ) && ( Tank.Node( 1 ).NewTemp > MaxTemp ) ) {
-						Event += NodeMass * Cp * ( MaxTemp - Tank.Node( 1 ).NewTemp );
-						Tank.Node( 1 ).NewTemp = MaxTemp;
+					if ( ( NodeNum == 1 ) && ( tank_node1_newTemp > MaxTemp ) ) {
+						Event += NodeMass * ( MaxTemp - tank_node1_newTemp );
+						tank_node1_newTemp = MaxTemp;
 					}
 				}
 
-				Esource += Qsource * dt;
-				Eloss += Qloss * dt;
-				Elosszone += Qlosszone * dt;
-				Eneeded += Qneeded * dt;
-				Eunmet += Qunmet * dt;
+				Esource += Qsource;
+				Eloss += Qloss;
+				Eneeded += Qneeded;
+				Eunmet += Qunmet;
 
 			} // NodeNum
 
-			Euse += UseMassFlowRate * Cp * (UseInletTemp - Tank.Node( Tank.UseOutletStratNode ).Temp) * dt;
+			Euse += UseMassFlowRate * ( UseInletTemp - tank_useOutletStratNode_temp );
+//? This is using the value of UseMassFlowRate from the last node
+//  If that is intended a comment would be a good idea
```

Comment:
@DeadParrot I don't think that's intended. It should be the use mass flow rate of the use outlet node. I'm going to create a unit test to exercise the bug and verify that this is happening.